source/util/bit_stream.cpp

```diff
@@ -0,0 +1,379 @@
+// Copyright (c) 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <algorithm>
+#include <cassert>
+#include <cstring>
+#include <sstream>
+#include <type_traits>
+
+#include "util/bit_stream.h"
+
+namespace spvutils {
+
+namespace {
+
+// Returns if the system is little-endian. Unfortunately only works during
+// runtime.
```

Comment:
I don't agree with what I see there. It is not a problem to cast
uint64_t* to either uint32_t* or uint16_t* or uint8_t*.

Casting uint8_t* to uint16_t* can be a problem, due to alignment. Casting structs or classes can be a problem due to padding and virtual function tables. But I don't see any problem with casting a pointer to a bigger uint to a pointer to a smaller uint.